packages/web/app/src/lib/preflight-sandbox/graphiql-plugin.tsx

```diff
@@ -0,0 +1,550 @@
+import {
+  CrossCircledIcon,
+  ExclamationTriangleIcon,
+  InfoCircledIcon,
+  Pencil1Icon,
+  TriangleRightIcon,
+} from '@radix-ui/react-icons';
+import { useParams } from '@tanstack/react-router';
+import type { LogMessage } from './execute-script';
+import PreflightWorker from './preflight-script-worker?worker';
+
+const PREFLIGHT_TIMEOUT = 30_000; // 30,000 ms = 30 seconds
+
+const usePreflightScriptStore = createWithEqualityFn(
+  persist<{
+    script: string;
+    env: string;
+    disabled: boolean;
+    setScript: Dispatch<string>;
+    setDisabled: Dispatch<boolean>;
+    setEnv: Dispatch<string | undefined>;
+  }>(
+    set => ({
+      script: '',
+      env: '',
+      disabled: false,
+      setScript: script => set({ script }),
+      setDisabled: disabled => set({ disabled }),
+      setEnv: env => set({ env }),
+    }),
+    {
+      name: 'preflight-script-storage',
+    },
+  ),
+  shallow,
+);
+
+const usePreflightScriptState = () =>
+  usePreflightScriptStore(state => ({
+    script: state.script,
+    env: state.env,
+    disabled: state.disabled,
+  }));
+const usePreflightScriptActions = () =>
+  usePreflightScriptStore(state => ({
+    setScript: state.setScript,
+    setEnv: state.setEnv,
+    setDisabled: state.setDisabled,
+  }));
+
+// We need to reassign to new instance after calling `.terminate()` method
+let preflightWorker = new PreflightWorker();
+
+export const preflightScriptPlugin: GraphiQLPlugin = {
+  icon: () => (
+    <svg
+      viewBox="0 0 256 256"
+      stroke="currentColor"
+      fill="none"
+      strokeLinecap="round"
+      strokeLinejoin="round"
+      strokeWidth="16"
+    >
+      <path d="M136 160h40" />
+      <path d="m80 96 40 32-40 32" />
+      <rect width="192" height="160" x="32" y="48" rx="8.5" />
+    </svg>
+  ),
+  title: 'Preflight Script',
+  content: PreflightScriptContent,
+};
+
+const classes = {
+  monaco: clsx('*:bg-[#10151f]'),
+  monacoMini: clsx('h-32 *:rounded-md *:bg-[#10151f]'),
+  icon: clsx('absolute -left-5 top-px'),
+};
+
+const sharedMonacoProps = {
+  theme: 'vs-dark',
+  className: classes.monaco,
+  options: {
+    minimap: { enabled: false },
+    padding: {
+      top: 10,
+    },
+    scrollbar: {
+      horizontalScrollbarSize: 6,
+      verticalScrollbarSize: 6,
+    },
+  },
+} satisfies ComponentProps<typeof MonacoEditor>;
+
+const monacoProps = {
+  env: {
+    ...sharedMonacoProps,
+    defaultLanguage: 'json',
+    options: {
+      ...sharedMonacoProps.options,
+      lineNumbers: 'off',
+      tabSize: 2,
+    },
+  },
+  script: {
+    ...sharedMonacoProps,
+    theme: 'vs-dark',
+    defaultLanguage: 'javascript',
+    options: {
+      ...sharedMonacoProps.options,
+    },
+  },
+} satisfies Record<'script' | 'env', ComponentProps<typeof MonacoEditor>>;
+
+type ConsoleLog = { type: 'log'; message: string };
+
+type PreflightScriptResult =
+  | { environmentVariables: Record<string, unknown> }
+  | ConsoleLog
+  | { error: Error };
+
```

Comment:
@dimaMachina Please add a type property so we don't do all these property in checks. Its a bit messy.

Reply:
added